### PR TITLE
Revert to use the old guava port cache (via jar)

### LINKED ISSUE #153: [BUG] Weigher API is not available in Store 4
**Describe the bug**
The Store 4 does not have the `Weigher` APIs as before in Cache APIs. This was an important part of the API since size of the cache is not always enough. Especially when the cache has values that are in variety of the size (byte), it is important to have a limit on the memory space. It allows us to keep small items and remove bigger ones when the cache is full. 

**To Reproduce**
Upgrade an already existing code with Store 3 to Store 4. 

**Expected behavior**
The `maximumWieght` function and `Weigher` interface should be available.

### PR DESCRIPTION
**Pull request by eyalgu**
This PR attempts to get the best of both worlds:
1) get the full functionality of guava's cache (including weigher)
2) don't compile any java code

This is achieved by depending on the old cache maven coordinates rather than building the new kotlin cache from source.

Given that this project in focused around repository pattern and does not care about caching in particular (and in fact our cache impl is very naive) I feel like abandoning the cache work makes sense.

closes #153

### CODE REVIEW

**Comment by eyalgu** [comment]
Also @ychescale9 (can't request review from you via UI)

**Comment by ychescale9** [comment]
@eyalgu It's sad but I guess it makes sense for this project as we're trying to ship 1.0 and it turns out that KMP really isn't the priority at all 😄 . Just a couple of comments:
- as far as I know the weigher API isn't the only requirement for **apollo-android** to be able to move to this cache implementation as they actually need a KMP implementation (@martinbonnin please correct me if this is not the case).
- Since our `MemoryPolicy` of Store only uses a small set of API surface from Guava cache. Does it make sense to ask user to bring their own implementation of the in-memory cache instead of bundling guava cache?

**Comment by martinbonnin** [comment]
Hi 👋 !

> as far as I know the weigher API isn't the only requirement for apollo-android to be able to move to this cache implementation as they actually need a KMP implementation 

That's correct. KMP might actually be a bigger requirement than the weigher API since we really want to make the apollo runtime KMP and that can be done initially without the weigher API while the opposite isn't true.

I'm might be missing some context here but it looks like the existing cache (https://github.com/dropbox/Store/tree/031d92e5b494ff6f6513182c8e638c9282778d50/cache/src/main/kotlin/com/dropbox/android/external/cache4) is already Kotlin?  How hard would it be to remove the JVM dependencies there? Maybe it would make sense to make that multiplatform and publish it as a separate repo where the weigher API can be added later? Unless there is already a good KMP in-memory lru cache implementation somewhere?

Also ping @tasomaniac would looked into this more than I did.

File: multicast/src/main/kotlin/com/dropbox/flow/multicast/ChannelManager.kt [hunk 1]
```diff
@@ -16,6 +16,7 @@
 @file:OptIn(ExperimentalStdlibApi::class)
 package com.dropbox.flow.multicast
 
+import com.dropbox.flow.multicast.ChannelManager.Message
```
Comment (yigit) [inline multicast/src/main/kotlin/com/dropbox/flow/multicast/ChannelManager.kt:19]: unnecessary import?

File: store/src/main/java/com/dropbox/android/external/store4/StoreBuilder.kt [hunk 2]
```diff
@@ -87,14 +88,15 @@ private class RealStoreBuilder<Key : Any, Input : Any, Output : Any>(
...
+    private var cachePolicy: MemoryPolicy<Key, Output>? = StoreDefaults.memoryPolicy
 
     override fun scope(scope: CoroutineScope): RealStoreBuilder<Key, Input, Output> {
         this.scope = scope
         return this
     }
 
-    override fun cachePolicy(memoryPolicy: MemoryPolicy?): RealStoreBuilder<Key, Input, Output> {
+    override fun cachePolicy(memoryPolicy: MemoryPolicy<Key, Output>?):
+        RealStoreBuilder<Key, Input, Output> {
```
Comment (yigit) [inline store/src/main/java/com/dropbox/android/external/store4/StoreBuilder.kt:99]: nit, if this doesnt' fit the previous line i think it is better to use multi line method description.
```suggestion
    override fun cachePolicy(
        memoryPolicy: MemoryPolicy<Key, Output>?
    ) : RealStoreBuilder<Key, Input, Output> {
```

**Comment by digitalbuddha** [comment]
Hi @martinbonnin , the current impl of the cache was our attempt to build a loading cache from the ground up. The effort was going okish until we noticed that weigher api was missing. The attempt at implementing weigher hit concurrency issues that are adding more complexity than we can take on. The decision was made to instead revert back to the binary of the full featured guava cache that we previously shaded.  As far as I know Apollo needs the weigher api to implement the cache as you are using it.

File: app/build.gradle [hunk 3]
```diff
@@ -67,7 +67,7 @@ dependencies {
     kapt(libraries.moshiCodegen)
     kapt(libraries.roomCompiler)
     implementation project(':store')
-    implementation project(':cache')
+    implementation 'com.nytimes.android:cache3:3.1.1'
```
Comment (digitalbuddha) [inline app/build.gradle:70]: just a note, that repo is archived. We might want to consider maintaining source at some point

File: multicast/src/main/kotlin/com/dropbox/flow/multicast/ChannelManager.kt (hunk 1 above)
Comment (digitalbuddha) [inline multicast/src/main/kotlin/com/dropbox/flow/multicast/ChannelManager.kt:19]:
```suggestion
```

File: store/src/main/java/com/dropbox/android/external/store4/MemoryPolicy.kt [hunk 4]
```diff
@@ -31,24 +47,30 @@ class MemoryPolicy internal constructor(
...
-    class MemoryPolicyBuilder {
+    val hasMaxWeight: Boolean = maxWeight != DEFAULT_SIZE_POLICY
+
+    class MemoryPolicyBuilder<Key : Any, Value : Any> {
         private var expireAfterWrite = DEFAULT_DURATION_POLICY
         private var expireAfterAccess = DEFAULT_DURATION_POLICY
-        private var maxSize: Long = -1
+        private var maxSize: Long = DEFAULT_SIZE_POLICY
+        private var maxWeight: Long = DEFAULT_SIZE_POLICY
+        private var weigher: Weigher<Key, Value> = OneWeigher
```
Comment (digitalbuddha) [inline store/src/main/java/com/dropbox/android/external/store4/MemoryPolicy.kt:57]: Store never had a memory policy for weigher, it was only an api in the cache artifact.  Seems fine to add

**Review by digitalbuddha — DISMISSED** [review]

**Comment by eyalgu** [comment]
I have an alternative approach we can take here:
* Dumb down the kotlin cache to be fully synchronized and remove loader support which store doesn't need
* Move the current store module to store-kmp and keep it depending on the kotlin cache. Add a cacheBuilder parameter to storeBuilder. Replace the newBuilder method with newBuilderNoCache
*  Add a new store module that add an extension method to builder - newBuilder which defaults to the guava cache

**Comment by ychescale9** [comment]
> I have an alternative approach we can take here:
> 
> * Dumb down the kotlin cache to be fully synchronized and remove loader support which store doesn't need
> * Move the current store module to store-kmp and keep it depending on the kotlin cache. Add a cacheBuilder parameter to storeBuilder. Replace the newBuilder method with newBuilderNoCache
> * Add a new store module that add an extension method to builder - newBuilder which defaults to the guava cache

Sounds like more complexity / overhead to maintain a separate module given the priority of KMP. But I'm happy with whatever you guys decide to do here :)

**Comment by digitalbuddha** [comment]
Past fixing checks, any reason not to merge this and call it a beta?

**Comment by digitalbuddha** [comment]
lgtm!

**Review by digitalbuddha — DISMISSED** [review]

**Review by digitalbuddha — DISMISSED** [review]

**Review by digitalbuddha — APPROVED** [review]

**Pull request merged**